# Preliminary Setup

In [1]:
%matplotlib inline
import os
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
os.chdir(r'D:/DS')
cwd = os.getcwd()
print(cwd)
warnings.filterwarnings('ignore')

D:\DS


# Problem Statement 1.1
Import the csv file of the stock of your choosing using 'pd.read_csv()' function into a dataframe. Shares of a company can be offered in more than one category. The category of a stock is indicated in the ‘Series’ column. If the csv file has data on more than one category, the ‘Date’ column will have repeating values. To avoid repetitions in the date, remove all the rows where 'Series' column is NOT 'EQ'. Analyze and understand each column properly. You'd find the head(), tail() and describe() functions to be immensely useful for exploration. You're free to carry out any other exploration of your own.

In [2]:
data = pd.read_csv(r'MINDTREE.csv')

In [3]:
data.shape

(494, 15)

In [4]:
data.head()

Symbol Series         Date  Prev Close  Open Price  High Price  \
0  MINDTREE     EQ  15-May-2017      510.60      507.35      515.00   
1  MINDTREE     EQ  16-May-2017      509.45      507.00      515.85   
2  MINDTREE     EQ  17-May-2017      511.30      507.05      519.70   
3  MINDTREE     EQ  18-May-2017      506.35      502.00      516.75   
4  MINDTREE     EQ  19-May-2017      492.90      491.95      502.95   

   Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
0     505.90      510.50       509.45         510.39                 271952   
1     506.00      511.00       511.30         512.69                 207011   
2     502.75      504.00       506.35         512.10                 391401   
3     486.00      487.50       492.90         505.60                 852336   
4     489.50      495.15       497.50         496.43                 352273   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
0  1.388005e+08           8261           100362                   36.90  
1  1.061317e+08           5550            82090                   39.65  
2  2.004375e+08          13341           150098                   38.35  
3  4.309377e+08          18513           170919                   20.05  
4  1.748782e+08           8326            57853                   16.42

In [5]:
data.tail()

Symbol Series         Date  Prev Close  Open Price  High Price  \
489  MINDTREE     EQ  07-May-2019      981.75       982.8      985.00   
490  MINDTREE     EQ  08-May-2019      979.95       981.0      985.00   
491  MINDTREE     EQ  09-May-2019      980.50       979.6      983.15   
492  MINDTREE     EQ  10-May-2019      980.45       980.4      982.90   
493  MINDTREE     EQ  13-May-2019      975.35       973.4      983.00   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
489     975.00       981.0       979.95         979.77                1242934   
490     971.05       982.0       980.50         979.84                4301316   
491     976.10       980.9       980.45         980.02                3787159   
492     973.00       974.0       975.35         977.55                 215170   
493     970.30       981.0       979.10         980.16                 485941   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
489  1.217784e+09          10864           884988                   71.20  
490  4.214603e+09          13038          3851210                   89.54  
491  3.711482e+09           8159          3479880                   91.89  
492  2.103396e+08           5128            26734                   12.42  
493  4.762997e+08          14005           222691                   45.83

In [6]:
data.describe()

Prev Close   Open Price   High Price    Low Price   Last Price  \
count   494.000000   494.000000   494.000000   494.000000   494.000000   
mean    778.486032   779.720547   792.118421   767.127328   779.634919   
std     214.140514   214.851383   218.056333   210.381329   214.484986   
min     442.500000   443.000000   448.750000   439.050000   441.400000   
25%     537.025000   537.975000   545.862500   528.012500   535.862500   
50%     835.075000   838.000000   851.525000   820.625000   835.025000   
75%     968.137500   968.500000   980.500000   949.450000   969.950000   
max    1161.950000  1165.000000  1183.800000  1147.800000  1166.000000   

       Close Price  Average Price  Total Traded Quantity      Turnover  \
count   494.000000     494.000000           4.940000e+02  4.940000e+02   
mean    779.434413     780.076619           1.388615e+06  1.191635e+09   
std     213.989087     214.038370           1.450169e+06  1.373025e+09   
min     442.500000     442.240000           4.630000e+04  2.305583e+07   
25%     537.400000     537.410000           5.724762e+05  3.399893e+08   
50%     835.875000     835.615000           1.006586e+06  8.580623e+08   
75%     969.775000     967.802500           1.631997e+06  1.431053e+09   
max    1161.950000    1167.480000           1.338521e+07  1.298866e+10   

       No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
count     494.000000     4.940000e+02               494.00000  
mean    33021.139676     3.463618e+05                29.18668  
std     30613.616442     3.420743e+05                14.05687  
min      1502.000000     1.505300e+04                 5.87000  
25%     14062.750000     1.523290e+05                18.69500  
50%     25609.500000     2.529855e+05                26.39500  
75%     42737.000000     4.366368e+05                37.06250  
max    311806.000000     3.851210e+06                91.89000

In [7]:
#Remove null values to avoid future errors
data.dropna(inplace = True)

In [8]:
#Checking type of Turnover Column since the values are displayed in exponential form with an 'e'
data.dtypes

Symbol                     object
Series                     object
Date                       object
Prev Close                float64
Open Price                float64
High Price                float64
Low Price                 float64
Last Price                float64
Close Price               float64
Average Price             float64
Total Traded Quantity       int64
Turnover                  float64
No. of Trades               int64
Deliverable Qty             int64
% Dly Qt to Traded Qty    float64
dtype: object

In [9]:
# Check for NOT 'EQ' values in Series column
data[data.Series != 'EQ']

Empty DataFrame
Columns: [Symbol, Series, Date, Prev Close, Open Price, High Price, Low Price, Last Price, Close Price, Average Price, Total Traded Quantity, Turnover, No. of Trades, Deliverable Qty, % Dly Qt to Traded Qty]
Index: []

#### Therefore, there are no rows where Series is not 'EQ'

In [10]:
data.apply(np.max)

Symbol                       MINDTREE
Series                             EQ
Date                      31-Oct-2018
Prev Close                    1161.95
Open Price                       1165
High Price                     1183.8
Low Price                      1147.8
Last Price                       1166
Close Price                   1161.95
Average Price                 1167.48
Total Traded Quantity        13385207
Turnover                  1.29887e+10
No. of Trades                  311806
Deliverable Qty               3851210
% Dly Qt to Traded Qty          91.89
dtype: object

In [11]:
data.apply(np.min)

Symbol                       MINDTREE
Series                             EQ
Date                      01-Apr-2019
Prev Close                      442.5
Open Price                        443
High Price                     448.75
Low Price                      439.05
Last Price                      441.4
Close Price                     442.5
Average Price                  442.24
Total Traded Quantity           46300
Turnover                  2.30558e+07
No. of Trades                    1502
Deliverable Qty                 15053
% Dly Qt to Traded Qty           5.87
dtype: object

# Problem Statement 1.2
Calculate the maximum, minimum and mean price for the last 90 days. (price=Closing Price unless stated otherwise) 

In [12]:
data.head()

Symbol Series         Date  Prev Close  Open Price  High Price  \
0  MINDTREE     EQ  15-May-2017      510.60      507.35      515.00   
1  MINDTREE     EQ  16-May-2017      509.45      507.00      515.85   
2  MINDTREE     EQ  17-May-2017      511.30      507.05      519.70   
3  MINDTREE     EQ  18-May-2017      506.35      502.00      516.75   
4  MINDTREE     EQ  19-May-2017      492.90      491.95      502.95   

   Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
0     505.90      510.50       509.45         510.39                 271952   
1     506.00      511.00       511.30         512.69                 207011   
2     502.75      504.00       506.35         512.10                 391401   
3     486.00      487.50       492.90         505.60                 852336   
4     489.50      495.15       497.50         496.43                 352273   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
0  1.388005e+08           8261           100362                   36.90  
1  1.061317e+08           5550            82090                   39.65  
2  2.004375e+08          13341           150098                   38.35  
3  4.309377e+08          18513           170919                   20.05  
4  1.748782e+08           8326            57853                   16.42

In [13]:
data.tail()

Symbol Series         Date  Prev Close  Open Price  High Price  \
489  MINDTREE     EQ  07-May-2019      981.75       982.8      985.00   
490  MINDTREE     EQ  08-May-2019      979.95       981.0      985.00   
491  MINDTREE     EQ  09-May-2019      980.50       979.6      983.15   
492  MINDTREE     EQ  10-May-2019      980.45       980.4      982.90   
493  MINDTREE     EQ  13-May-2019      975.35       973.4      983.00   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
489     975.00       981.0       979.95         979.77                1242934   
490     971.05       982.0       980.50         979.84                4301316   
491     976.10       980.9       980.45         980.02                3787159   
492     973.00       974.0       975.35         977.55                 215170   
493     970.30       981.0       979.10         980.16                 485941   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
489  1.217784e+09          10864           884988                   71.20  
490  4.214603e+09          13038          3851210                   89.54  
491  3.711482e+09           8159          3479880                   91.89  
492  2.103396e+08           5128            26734                   12.42  
493  4.762997e+08          14005           222691                   45.83

In [14]:
data['Close Price'].max()

1161.95

In [15]:
data['Close Price'].min()

442.5

In [16]:
data['Close Price'].mean()

779.4344129554655

#### The above (max, min, mean) was for all the dates under consideration.
#### For only past 90 days we have modify the Date values

In [17]:
data['Date'].dtype

dtype('O')

In [18]:
data['Index'] = data.index
data.tail()

Symbol Series         Date  Prev Close  Open Price  High Price  \
489  MINDTREE     EQ  07-May-2019      981.75       982.8      985.00   
490  MINDTREE     EQ  08-May-2019      979.95       981.0      985.00   
491  MINDTREE     EQ  09-May-2019      980.50       979.6      983.15   
492  MINDTREE     EQ  10-May-2019      980.45       980.4      982.90   
493  MINDTREE     EQ  13-May-2019      975.35       973.4      983.00   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
489     975.00       981.0       979.95         979.77                1242934   
490     971.05       982.0       980.50         979.84                4301316   
491     976.10       980.9       980.45         980.02                3787159   
492     973.00       974.0       975.35         977.55                 215170   
493     970.30       981.0       979.10         980.16                 485941   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
489  1.217784e+09          10864           884988                   71.20   
490  4.214603e+09          13038          3851210                   89.54   
491  3.711482e+09           8159          3479880                   91.89   
492  2.103396e+08           5128            26734                   12.42   
493  4.762997e+08          14005           222691                   45.83   

     Index  
489    489  
490    490  
491    491  
492    492  
493    493

In [19]:
# create a separate dataframe only for past 90 days
data90 = data.loc[(data['Index'] >= 403) & (data['Index'] <= 493)]

In [20]:
data90.head()

Symbol Series         Date  Prev Close  Open Price  High Price  \
403  MINDTREE     EQ  28-Dec-2018      861.85      863.00       869.0   
404  MINDTREE     EQ  31-Dec-2018      864.10      870.00       871.7   
405  MINDTREE     EQ  01-Jan-2019      864.50      864.00       867.1   
406  MINDTREE     EQ  02-Jan-2019      863.40      858.00       878.0   
407  MINDTREE     EQ  03-Jan-2019      860.50      861.15       865.0   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
403     858.05      865.20        864.1         863.70                 420544   
404     861.55      864.00        864.5         866.31                 317152   
405     853.40      863.15        863.4         861.25                 399996   
406     857.20      859.00        860.5         870.52                 892157   
407     824.35      834.30        834.8         848.07                 908897   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
403  3.632257e+08           7846            49437                   11.76   
404  2.747529e+08          11757            61520                   19.40   
405  3.444982e+08           7996            27003                    6.75   
406  7.766405e+08          19259            84733                    9.50   
407  7.708068e+08          53508           305402                   33.60   

     Index  
403    403  
404    404  
405    405  
406    406  
407    407

In [21]:
data90.tail()

Symbol Series         Date  Prev Close  Open Price  High Price  \
489  MINDTREE     EQ  07-May-2019      981.75       982.8      985.00   
490  MINDTREE     EQ  08-May-2019      979.95       981.0      985.00   
491  MINDTREE     EQ  09-May-2019      980.50       979.6      983.15   
492  MINDTREE     EQ  10-May-2019      980.45       980.4      982.90   
493  MINDTREE     EQ  13-May-2019      975.35       973.4      983.00   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
489     975.00       981.0       979.95         979.77                1242934   
490     971.05       982.0       980.50         979.84                4301316   
491     976.10       980.9       980.45         980.02                3787159   
492     973.00       974.0       975.35         977.55                 215170   
493     970.30       981.0       979.10         980.16                 485941   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
489  1.217784e+09          10864           884988                   71.20   
490  4.214603e+09          13038          3851210                   89.54   
491  3.711482e+09           8159          3479880                   91.89   
492  2.103396e+08           5128            26734                   12.42   
493  4.762997e+08          14005           222691                   45.83   

     Index  
489    489  
490    490  
491    491  
492    492  
493    493

#### Now, we have a subset of the original data only for the past 90 days

In [22]:
data90['Close Price'].max()

989.4

In [23]:
data90['Close Price'].min()

812.25

In [24]:
data90['Close Price'].mean()

918.0131868131863

# Problem Statement 1.3
Analyse the data types for each column of the dataframe. Pandas knows how to deal with dates in an intelligent manner. But to make use of Pandas functionality for dates, you need to ensure that the column is of type 'datetime64(ns)'. Change the date column from 'object' type to 'datetime64(ns)' for future convenience. See what happens if you subtract the minimum value of the date column from the maximum value.

In [25]:
data['Date'].dtype

dtype('O')

In [26]:
from datetime import datetime
data['Date'] = data['Date'].astype(datetime)

In [27]:
data['Date'].dtype

dtype('O')

In [28]:
data['Date'] = pd.to_datetime(data['Date'], errors='coerce')

In [29]:
data['Date'].dtype

dtype('<M8[ns]')

In [30]:
np.dtype('datetime64[ns]') == np.dtype('<M8[ns]')

True

In [31]:
data['Date'].max() - data['Date'].min()

Timedelta('728 days 00:00:00')

# Problem Statement 1.4
In a separate array , calculate the monthwise VWAP (Volume Weighted Average Price ) of the stock.
( VWAP = sum(price*volume)/sum(volume) )
To know more about VWAP , visit - VWAP definition
{Hint : Create a new dataframe column ‘Month’. The values for this column can be derived from the ‘Date” column by using appropriate pandas functions. Similarly, create a column ‘Year’ and initialize it. Then use the 'groupby()' function by month and year. Finally, calculate the vwap value for each month (i.e. for each group created). 

Key Takeaways regarding VWAPS:

1. The volume weighted average price (VWAP) appears as a single line on intraday charts (1 minute, 15 minute, and so on), similar to how a moving average looks.

2. A rising VWAP, and/or the price above the VWAP line, means the price is likely in an uptrend.

3. A declining VWAP, and/or the price below the VWAP line, means the price is likely in a downtrend.

4. Don't rely on VWAP exclusively to determine trend, since it is only showing a historical average, and not what is happening currently or in the future.

5. Investors may use VWAP to assess the price they paid for a security throughout the day. At the end of the day, if the price they bought at is higher than the VWAP, then they may have overpaid. If it is less than VWAP, then they purchased shares at a good price for that day.

Formula to calculate VWAP:

VWAP is calculated by adding up the dollars traded for every transaction (price multiplied by the number of shares traded) and then dividing by the total shares traded.

In [32]:
data['Date'].max()

Timestamp('2019-05-13 00:00:00')

In [33]:
data['Month'] = pd.DatetimeIndex(data['Date']).month
data['Year'] = pd.DatetimeIndex(data['Date']).year

In [34]:
data.head(10)

Symbol Series       Date  Prev Close  Open Price  High Price  Low Price  \
0  MINDTREE     EQ 2017-05-15      510.60      507.35      515.00     505.90   
1  MINDTREE     EQ 2017-05-16      509.45      507.00      515.85     506.00   
2  MINDTREE     EQ 2017-05-17      511.30      507.05      519.70     502.75   
3  MINDTREE     EQ 2017-05-18      506.35      502.00      516.75     486.00   
4  MINDTREE     EQ 2017-05-19      492.90      491.95      502.95     489.50   
5  MINDTREE     EQ 2017-05-22      497.50      499.95      511.85     497.10   
6  MINDTREE     EQ 2017-05-23      502.35      504.00      507.45     491.00   
7  MINDTREE     EQ 2017-05-24      501.15      502.00      515.00     494.05   
8  MINDTREE     EQ 2017-05-25      509.95      513.70      549.50     508.45   
9  MINDTREE     EQ 2017-05-26      543.90      540.00      544.00     532.00   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
0      510.50       509.45         510.39                 271952   
1      511.00       511.30         512.69                 207011   
2      504.00       506.35         512.10                 391401   
3      487.50       492.90         505.60                 852336   
4      495.15       497.50         496.43                 352273   
5      501.55       502.35         504.78                 399974   
6      498.00       501.15         498.72                 452893   
7      511.95       509.95         506.72                 823967   
8      542.50       543.90         532.14                1632427   
9      533.45       535.20         537.15                 461613   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
0  1.388005e+08           8261           100362                   36.90   
1  1.061317e+08           5550            82090                   39.65   
2  2.004375e+08          13341           150098                   38.35   
3  4.309377e+08          18513           170919                   20.05   
4  1.748782e+08           8326            57853                   16.42   
5  2.018999e+08           9779           107089                   26.77   
6  2.258664e+08           8567           198703                   43.87   
7  4.175211e+08          15083           351179                   42.62   
8  8.686829e+08          25072           693340                   42.47   
9  2.479538e+08          12606           113087                   24.50   

   Index  Month  Year  
0      0      5  2017  
1      1      5  2017  
2      2      5  2017  
3      3      5  2017  
4      4      5  2017  
5      5      5  2017  
6      6      5  2017  
7      7      5  2017  
8      8      5  2017  
9      9      5  2017

In [35]:
data.tail()

Symbol Series       Date  Prev Close  Open Price  High Price  \
489  MINDTREE     EQ 2019-05-07      981.75       982.8      985.00   
490  MINDTREE     EQ 2019-05-08      979.95       981.0      985.00   
491  MINDTREE     EQ 2019-05-09      980.50       979.6      983.15   
492  MINDTREE     EQ 2019-05-10      980.45       980.4      982.90   
493  MINDTREE     EQ 2019-05-13      975.35       973.4      983.00   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
489     975.00       981.0       979.95         979.77                1242934   
490     971.05       982.0       980.50         979.84                4301316   
491     976.10       980.9       980.45         980.02                3787159   
492     973.00       974.0       975.35         977.55                 215170   
493     970.30       981.0       979.10         980.16                 485941   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
489  1.217784e+09          10864           884988                   71.20   
490  4.214603e+09          13038          3851210                   89.54   
491  3.711482e+09           8159          3479880                   91.89   
492  2.103396e+08           5128            26734                   12.42   
493  4.762997e+08          14005           222691                   45.83   

     Index  Month  Year  
489    489      5  2019  
490    490      5  2019  
491    491      5  2019  
492    492      5  2019  
493    493      5  2019

#### Now we have separate columns for Year and Month

In [36]:
data['Month'].min()

1

In [37]:
data['Month'].max()

12

In [38]:
data['Year'].min()

2017

In [39]:
data['Year'].max()

2019

#### Therefore, the entire dataset is divided into 1st(January) to 12th(December) month and 2017 and 2019.

In [40]:
data_month = data.groupby(['Month'])
data_year = data.groupby(['Year'])

In [41]:
data_month.head(1)

Symbol Series       Date  Prev Close  Open Price  High Price  \
0    MINDTREE     EQ 2017-05-15      510.60      507.35      515.00   
13   MINDTREE     EQ 2017-06-01      543.95      545.05      549.00   
34   MINDTREE     EQ 2017-07-03      529.00      530.90      534.00   
55   MINDTREE     EQ 2017-08-01      477.45      477.35      479.20   
76   MINDTREE     EQ 2017-09-01      461.00      463.30      466.60   
97   MINDTREE     EQ 2017-10-03      464.85      465.00      482.05   
117  MINDTREE     EQ 2017-11-01      479.65      483.30      484.00   
139  MINDTREE     EQ 2017-12-01      539.20      539.45      547.60   
159  MINDTREE     EQ 2018-01-01      612.40      610.80      618.35   
181  MINDTREE     EQ 2018-02-01      770.00      765.55      769.00   
200  MINDTREE     EQ 2018-03-01      815.10      820.00      824.35   
219  MINDTREE     EQ 2018-04-02      771.95      782.00      812.80   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
0       505.90      510.50       509.45         510.39                 271952   
13      535.00      538.00       537.40         540.08                 504801   
34      525.10      529.00       529.05         528.71                 823697   
55      469.45      472.35       472.60         474.85                 371419   
76      459.30      460.00       460.85         462.28                 219330   
97      465.00      480.05       479.20         478.36                 518751   
117     471.00      473.10       474.75         476.51                 328658   
139     527.45      530.50       529.90         539.50                 277028   
159     605.30      606.05       606.90         610.63                 318744   
181     745.05      757.75       753.60         757.46                 936000   
200     810.05      812.00       814.05         818.21                1343617   
219     772.50      811.00       810.20         802.53                2650454   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
0    1.388005e+08           8261           100362                   36.90   
13   2.726343e+08          18409           182246                   36.10   
34   4.354947e+08          17213           486031                   59.01   
55   1.763685e+08          13465           183546                   49.42   
76   1.013926e+08           7141            43567                   19.86   
97   2.481503e+08          10480           139459                   26.88   
117  1.566100e+08          12038           198973                   60.54   
139  1.494555e+08           8714            56836                   20.52   
159  1.946341e+08           6312           107284                   33.66   
181  7.089838e+08          42431           352018                   37.61   
200  1.099362e+09          25412           137094                   10.20   
219  2.127057e+09          72612           679665                   25.64   

     Index  Month  Year  
0        0      5  2017  
13      13      6  2017  
34      34      7  2017  
55      55      8  2017  
76      76      9  2017  
97      97     10  2017  
117    117     11  2017  
139    139     12  2017  
159    159      1  2018  
181    181      2  2018  
200    200      3  2018  
219    219      4  2018

#### Therefore, the data starts from May 2017 and goes on to 2019
#### Entries from 219 onto 494 are from May 2018 to 2019

In [42]:
data_year.head(1)

Symbol Series       Date  Prev Close  Open Price  High Price  \
0    MINDTREE     EQ 2017-05-15       510.6      507.35      515.00   
159  MINDTREE     EQ 2018-01-01       612.4      610.80      618.35   
405  MINDTREE     EQ 2019-01-01       864.5      864.00      867.10   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
0        505.9      510.50       509.45         510.39                 271952   
159      605.3      606.05       606.90         610.63                 318744   
405      853.4      863.15       863.40         861.25                 399996   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
0    1.388005e+08           8261           100362                   36.90   
159  1.946341e+08           6312           107284                   33.66   
405  3.444982e+08           7996            27003                    6.75   

     Index  Month  Year  
0        0      5  2017  
159    159      1  2018  
405    405      1  2019

#### Entries from 0 - 158 : 2017
#### Entries from 159 - 404 : 2018
#### Entries rom 405 - 494 : 2019

In [43]:
data_month.mean()

Prev Close  Open Price  High Price   Low Price  Last Price  \
Month                                                               
1      764.577778  766.562222  782.050000  754.606667  768.474444   
2      818.662821  818.225641  832.753846  805.462821  820.478205   
3      869.611842  872.440789  884.298684  858.665789  869.925000   
4      923.841250  927.175000  943.592500  917.987500  933.195000   
5      860.183721  862.483721  873.510465  849.996512  859.406977   
6      764.789286  767.298810  775.370238  751.958333  764.352381   
7      762.861628  763.955814  773.380233  749.638372  760.424419   
8      732.083333  732.697619  744.997619  724.188095  735.947619   
9      764.624359  766.255128  776.639744  751.020513  764.087179   
10     714.117073  711.752439  726.520732  695.232927  709.947561   
11     663.884524  665.002381  676.779762  654.719048  666.007143   
12     714.936250  715.086250  727.925000  704.168750  715.745000   

       Close Price  Average Price  Total Traded Quantity      Turnover  \
Month                                                                    
1       768.736667     769.272222           2.219573e+06  1.787288e+09   
2       820.197436     820.471282           1.418572e+06  1.161449e+09   
3       869.415789     870.647632           2.049464e+06  1.778634e+09   
4       932.576250     932.708000           2.333345e+06  2.198426e+09   
5       859.126744     861.873023           1.261509e+06  1.190994e+09   
6       763.932143     763.879286           8.900523e+05  7.360172e+08   
7       760.393023     760.892791           1.539243e+06  1.377818e+09   
8       736.034524     735.616905           7.700193e+05  7.130137e+08   
9       762.634615     764.126667           1.009829e+06  9.721465e+08   
10      710.029268     710.670488           1.259713e+06  9.967849e+08   
11      665.916667     666.079286           9.150853e+05  6.581350e+08   
12      716.462500     716.944000           9.984260e+05  7.419116e+08   

       No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty       Index  \
Month                                                                       
1       48113.888889    394758.600000               22.387778  295.488889   
2       38064.205128    359528.820513               26.994615  316.923077   
3       47264.526316    488773.789474               24.653684  333.000000   
4       47427.500000    520243.075000               31.579500  346.325000   
5       25457.488372    475256.465116               33.163023  221.046512   
6       22845.857143    241547.404762               28.539048  147.500000   
7       35587.069767    420226.720930               35.134419  171.651163   
8       20114.761905    183181.452381               29.248571  190.000000   
9       24894.641026    247718.692308               31.674359  200.692308   
10      34284.804878    342779.268293               31.934634  233.268293   
11      26171.333333    266336.071429               31.929286  245.119048   
12      26615.900000    217209.325000               22.711000  271.500000   

              Year  
Month               
1      2018.511111  
2      2018.512821  
3      2018.500000  
4      2018.475000  
5      2017.883721  
6      2017.500000  
7      2017.511628  
8      2017.500000  
9      2017.461538  
10     2017.512195  
11     2017.476190  
12     2017.500000

In [44]:
data_year.mean()

Prev Close  Open Price  High Price   Low Price  Last Price  Close Price  \
Year                                                                            
2017  504.475157  504.951572  512.437421  497.979245  504.819182   505.115409   
2018  905.140447  906.657927  922.840244  889.942073  906.680894   906.165244   
2019  917.932584  919.739326  930.452247  908.499438  919.437079   919.220225   

      Average Price  Total Traded Quantity      Turnover  No. of Trades  \
Year                                                                      
2017     505.836415           5.808072e+05  3.001232e+08   13549.641509   
2018     906.759797           1.804320e+06  1.643528e+09   43881.227642   
2019     919.853371           1.682753e+06  1.535284e+09   37789.528090   

      Deliverable Qty  % Dly Qt to Traded Qty  Index     Month  
Year                                                            
2017    219460.974843               37.950629   79.0  8.660377  
2018    384979.085366               23.272033  281.5  6.467480  
2019    466332.516854               29.878090  449.0  2.651685

In [45]:
data_month['Close Price'].mean()

Month
1     768.736667
2     820.197436
3     869.415789
4     932.576250
5     859.126744
6     763.932143
7     760.393023
8     736.034524
9     762.634615
10    710.029268
11    665.916667
12    716.462500
Name: Close Price, dtype: float64

In [46]:
data_year['Close Price'].mean()

Year
2017    505.115409
2018    906.165244
2019    919.220225
Name: Close Price, dtype: float64

#### Create new column for product of Close Price and Total Traded Quantity (ie. Volume)

In [47]:
#data['VWAP'] = (data.groupby('Month')['Close Price'].mean() * data.groupby('Month')['Total Traded Quantity'].mean())

In [48]:
"""def vwap(df):
    q = df['Total Traded Quantity']
    p = df['Close Price']
    return data.assign(vwap =(p * q).cumsum() / q.cumsum())

data.groupby(data['Month'], group_keys = False).apply(vwap)"""

"def vwap(df):\n    q = df['Total Traded Quantity']\n    p = df['Close Price']\n    return data.assign(vwap =(p * q).cumsum() / q.cumsum())\n\ndata.groupby(data['Month'], group_keys = False).apply(vwap)"

In [49]:
#below wrong, use cumsum
#groupby_month['VWAP'] = (groupby_month['Close Price']*groupby_month['Total Traded Quantity'])/groupby_month['Total Traded Quantity']

In [50]:
#groupby_month.sort_values(['Month'], axis=0, ascending=True, inplace=True) 

In [51]:
data_month

In [52]:
data_jan = data_month.get_group(1)

In [53]:
data_jan.tail()

Symbol Series       Date  Prev Close  Open Price  High Price  \
423  MINDTREE     EQ 2019-01-25      910.40      916.05      925.00   
424  MINDTREE     EQ 2019-01-28      886.10      872.00      884.25   
425  MINDTREE     EQ 2019-01-29      868.60      870.00      911.00   
426  MINDTREE     EQ 2019-01-30      903.60      909.20      919.00   
427  MINDTREE     EQ 2019-01-31      890.45      900.00      902.00   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
423     880.80      887.95       886.10         903.83                2686280   
424     856.05      868.15       868.60         870.47                2225129   
425     866.00      907.35       903.60         890.49                1990949   
426     885.10      890.00       890.45         902.76                1661063   
427     880.00      893.90       894.05         892.33                1211981   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
423  2.427941e+09          49029           317435                   11.82   
424  1.936909e+09          53270           322191                   14.48   
425  1.772915e+09          46222           283796                   14.25   
426  1.499536e+09          50115           309584                   18.64   
427  1.081483e+09          26225           146502                   12.09   

     Index  Month  Year  
423    423      1  2019  
424    424      1  2019  
425    425      1  2019  
426    426      1  2019  
427    427      1  2019

In [54]:
#data_jan['Price_Vol'] = data_jan.apply(lambda row: (row['Close Price']*row['Total Traded Quantity'], axis = 1))

In [55]:
data_jan.dtypes

Symbol                            object
Series                            object
Date                      datetime64[ns]
Prev Close                       float64
Open Price                       float64
High Price                       float64
Low Price                        float64
Last Price                       float64
Close Price                      float64
Average Price                    float64
Total Traded Quantity              int64
Turnover                         float64
No. of Trades                      int64
Deliverable Qty                    int64
% Dly Qt to Traded Qty           float64
Index                              int64
Month                              int64
Year                               int64
dtype: object

In [56]:
data_jan['Price_Vol'] = data_jan['Close Price'] * data_jan['Total Traded Quantity']

In [57]:
data_jan.head()

Symbol Series       Date  Prev Close  Open Price  High Price  \
159  MINDTREE     EQ 2018-01-01      612.40      610.80      618.35   
160  MINDTREE     EQ 2018-01-02      606.90      603.05      611.35   
161  MINDTREE     EQ 2018-01-03      605.05      606.00      613.80   
162  MINDTREE     EQ 2018-01-04      599.30      601.00      605.75   
163  MINDTREE     EQ 2018-01-05      599.90      600.50      605.00   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
159     605.30      606.05       606.90         610.63                 318744   
160     598.00      605.05       605.05         604.00                 528274   
161     595.70      599.50       599.30         606.68                 853218   
162     593.05      599.45       599.90         599.41                1122243   
163     597.00      602.00       602.30         600.57                 908732   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
159  1.946341e+08           6312           107284                   33.66   
160  3.190759e+08          10102           110148                   20.85   
161  5.176280e+08          13411           276474                   32.40   
162  6.726834e+08          19195           651722                   58.07   
163  5.457605e+08          12565           579769                   63.80   

     Index  Month  Year    Price_Vol  
159    159      1  2018  193445733.6  
160    160      1  2018  319632183.7  
161    161      1  2018  511333547.4  
162    162      1  2018  673233575.7  
163    163      1  2018  547329283.6

In [58]:
# Using cumulative sum to calculate VWAP for each day in January
data_jan['VWAP'] = data_jan['Price_Vol'].sum()/data_jan['Total Traded Quantity'].sum()

In [59]:
data_jan.head()

Symbol Series       Date  Prev Close  Open Price  High Price  \
159  MINDTREE     EQ 2018-01-01      612.40      610.80      618.35   
160  MINDTREE     EQ 2018-01-02      606.90      603.05      611.35   
161  MINDTREE     EQ 2018-01-03      605.05      606.00      613.80   
162  MINDTREE     EQ 2018-01-04      599.30      601.00      605.75   
163  MINDTREE     EQ 2018-01-05      599.90      600.50      605.00   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
159     605.30      606.05       606.90         610.63                 318744   
160     598.00      605.05       605.05         604.00                 528274   
161     595.70      599.50       599.30         606.68                 853218   
162     593.05      599.45       599.90         599.41                1122243   
163     597.00      602.00       602.30         600.57                 908732   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
159  1.946341e+08           6312           107284                   33.66   
160  3.190759e+08          10102           110148                   20.85   
161  5.176280e+08          13411           276474                   32.40   
162  6.726834e+08          19195           651722                   58.07   
163  5.457605e+08          12565           579769                   63.80   

     Index  Month  Year    Price_Vol        VWAP  
159    159      1  2018  193445733.6  806.352228  
160    160      1  2018  319632183.7  806.352228  
161    161      1  2018  511333547.4  806.352228  
162    162      1  2018  673233575.7  806.352228  
163    163      1  2018  547329283.6  806.352228

#### Approach used in 1.4 : Groupby month in main dataframe and then create new data frame for each month and calculate VWAP for each month individually. January is done in said method. Similar approach can be followed fro the other months.

# Problem Statement 1.5
Write a function to calculate the average price over the last N days of the stock price data where N is a user defined parameter. Write a second function to calculate the profit/loss percentage over the last N days.
Calculate the average price AND the profit/loss percentages over the course of last -
1 week, 2 weeks, 1 month, 3 months, 6 months and 1 year.
{Note : Profit/Loss percentage between N days is the percentage change between the closing prices of the 2 days }

In [60]:
data.tail()

Symbol Series       Date  Prev Close  Open Price  High Price  \
489  MINDTREE     EQ 2019-05-07      981.75       982.8      985.00   
490  MINDTREE     EQ 2019-05-08      979.95       981.0      985.00   
491  MINDTREE     EQ 2019-05-09      980.50       979.6      983.15   
492  MINDTREE     EQ 2019-05-10      980.45       980.4      982.90   
493  MINDTREE     EQ 2019-05-13      975.35       973.4      983.00   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
489     975.00       981.0       979.95         979.77                1242934   
490     971.05       982.0       980.50         979.84                4301316   
491     976.10       980.9       980.45         980.02                3787159   
492     973.00       974.0       975.35         977.55                 215170   
493     970.30       981.0       979.10         980.16                 485941   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
489  1.217784e+09          10864           884988                   71.20   
490  4.214603e+09          13038          3851210                   89.54   
491  3.711482e+09           8159          3479880                   91.89   
492  2.103396e+08           5128            26734                   12.42   
493  4.762997e+08          14005           222691                   45.83   

     Index  Month  Year  
489    489      5  2019  
490    490      5  2019  
491    491      5  2019  
492    492      5  2019  
493    493      5  2019

In [61]:
n = input('Period of time to perform calculation(in days): ')
n = int(n)
# extracting the data for the user given number of days
data_n = data.loc[(data['Index'] >= 493 - n) & (data['Index'] <= 493)]

def avg_price(n):   
    ap = data_n['Close Price'].sum()/n
    print("Average price for the given time period: ", ap)

    
def pl_percentage(n):
    pl_df = data_n.iloc[[0, -1]]
    pl_df.head()
    a = pl_df['Close Price']
    a =a.pct_change()
    a = a.iloc[-1] * 100
    print("PL Percentage change for the given time period: ", a, "%")
    
avg_price(n)
pl_percentage(n)   

Period of time to perform calculation(in days): 7
Average price for the given time period:  1118.3857142857144
PL Percentage change for the given time period:  -0.2343590788669192 %


Period of time to perform calculation(in days): 7
Average price for the given time period:  1118.3857142857144
PL Percentage change for the given time period:  -0.2343590788669192 %


Period of time to perform calculation(in days): 14
Average price for the given time period:  1048.1928571428573
PL Percentage change for the given time period:  1.0944759938048554 %


Period of time to perform calculation(in days): 30
Average price for the given time period:  999.3366666666665
PL Percentage change for the given time period:  3.008942661756975 %


Period of time to perform calculation(in days): 90
Average price for the given time period:  928.2133333333335
PL Percentage change for the given time period:  13.308644832773986 %


Period of time to perform calculation(in days): 180
Average price for the given time period:  937.8138888888889
PL Percentage change for the given time period:  -0.38661104893681353 %


Period of time to perform calculation(in days): 365
Average price for the given time period:  881.9361643835617
PL Percentage change for the given time period:  91.67971808927173 %

# Problem Statement 1.6
 Add a column 'Day_Perc_Change' where the values are the daily change in percentages i.e. the percentage change between 2 consecutive day's closing prices. Instead of using the basic mathematical formula for computing the same, use 'pct_change()' function provided by Pandas for dataframes. You will note that the first entry of the column will have a ‘Nan’ value. Why does this happen? Either remove the first row, or set the entry to 0 before proceeding.
 

In [62]:
data.head()

Symbol Series       Date  Prev Close  Open Price  High Price  Low Price  \
0  MINDTREE     EQ 2017-05-15      510.60      507.35      515.00     505.90   
1  MINDTREE     EQ 2017-05-16      509.45      507.00      515.85     506.00   
2  MINDTREE     EQ 2017-05-17      511.30      507.05      519.70     502.75   
3  MINDTREE     EQ 2017-05-18      506.35      502.00      516.75     486.00   
4  MINDTREE     EQ 2017-05-19      492.90      491.95      502.95     489.50   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
0      510.50       509.45         510.39                 271952   
1      511.00       511.30         512.69                 207011   
2      504.00       506.35         512.10                 391401   
3      487.50       492.90         505.60                 852336   
4      495.15       497.50         496.43                 352273   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
0  1.388005e+08           8261           100362                   36.90   
1  1.061317e+08           5550            82090                   39.65   
2  2.004375e+08          13341           150098                   38.35   
3  4.309377e+08          18513           170919                   20.05   
4  1.748782e+08           8326            57853                   16.42   

   Index  Month  Year  
0      0      5  2017  
1      1      5  2017  
2      2      5  2017  
3      3      5  2017  
4      4      5  2017

#### Calculating % Change using pct_change() function

In [63]:
data['Day_Perc_Change'] = data[['Close Price']].pct_change()*100
data.head()

Symbol Series       Date  Prev Close  Open Price  High Price  Low Price  \
0  MINDTREE     EQ 2017-05-15      510.60      507.35      515.00     505.90   
1  MINDTREE     EQ 2017-05-16      509.45      507.00      515.85     506.00   
2  MINDTREE     EQ 2017-05-17      511.30      507.05      519.70     502.75   
3  MINDTREE     EQ 2017-05-18      506.35      502.00      516.75     486.00   
4  MINDTREE     EQ 2017-05-19      492.90      491.95      502.95     489.50   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
0      510.50       509.45         510.39                 271952   
1      511.00       511.30         512.69                 207011   
2      504.00       506.35         512.10                 391401   
3      487.50       492.90         505.60                 852336   
4      495.15       497.50         496.43                 352273   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
0  1.388005e+08           8261           100362                   36.90   
1  1.061317e+08           5550            82090                   39.65   
2  2.004375e+08          13341           150098                   38.35   
3  4.309377e+08          18513           170919                   20.05   
4  1.748782e+08           8326            57853                   16.42   

   Index  Month  Year  Day_Perc_Change  
0      0      5  2017              NaN  
1      1      5  2017         0.363137  
2      2      5  2017        -0.968120  
3      3      5  2017        -2.656265  
4      4      5  2017         0.933252

#### Removing first row to avoid NaN Value

In [64]:
data = data.iloc[1:]
data.head()

Symbol Series       Date  Prev Close  Open Price  High Price  Low Price  \
1  MINDTREE     EQ 2017-05-16      509.45      507.00      515.85     506.00   
2  MINDTREE     EQ 2017-05-17      511.30      507.05      519.70     502.75   
3  MINDTREE     EQ 2017-05-18      506.35      502.00      516.75     486.00   
4  MINDTREE     EQ 2017-05-19      492.90      491.95      502.95     489.50   
5  MINDTREE     EQ 2017-05-22      497.50      499.95      511.85     497.10   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
1      511.00       511.30         512.69                 207011   
2      504.00       506.35         512.10                 391401   
3      487.50       492.90         505.60                 852336   
4      495.15       497.50         496.43                 352273   
5      501.55       502.35         504.78                 399974   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
1  1.061317e+08           5550            82090                   39.65   
2  2.004375e+08          13341           150098                   38.35   
3  4.309377e+08          18513           170919                   20.05   
4  1.748782e+08           8326            57853                   16.42   
5  2.018999e+08           9779           107089                   26.77   

   Index  Month  Year  Day_Perc_Change  
1      1      5  2017         0.363137  
2      2      5  2017        -0.968120  
3      3      5  2017        -2.656265  
4      4      5  2017         0.933252  
5      5      5  2017         0.974874

# Problem Statement 1.7

Add another column 'Trend' whose values are:

    'Slight or No change' for 'Day_Perc_Change' in between -0.5 and 0.5
    'Slight positive' for 'Day_Perc_Change' in between 0.5 and 1
    'Slight negative' for 'Day_Perc_Change' in between -0.5 and -1
    'Positive' for 'Day_Perc_Change' in between 1 and 3
    'Negative' for 'Day_Perc_Change' in between -1 and -3
    'Among top gainers' for 'Day_Perc_Change' in between 3 and 7
    'Among top losers' for 'Day_Perc_Change' in between -3 and -7
    'Bull run' for 'Day_Perc_Change' >7
    'Bear drop' for 'Day_Perc_Change' <-7



In [71]:
data['Trend'] = np.where(data['Day_Perc_Change']>= -0.5, 'Slight or No change', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']<= 0.5, 'Slight or No change', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']>= 0.5, 'Slight Positive', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']<= 1, 'Slight Positive', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']<= -0.5, 'Slight Negative', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']>= -1, 'Slight Negative', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']>= 1, 'Positive', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']<= 3, 'Positive', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']<= -1, 'Negative', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']>= -3, 'Negative', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']>= 3, 'Among top gainers', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']<= 7, 'Among top gainers', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']<= -3, 'Among top losers', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']>= -7, 'Among top losers', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']> 7, 'Bull run', 'NA')
data['Trend'] = np.where(data['Day_Perc_Change']< -7, 'Bear Drop', 'NA')

In [72]:
data.head()

Symbol Series       Date  Prev Close  Open Price  High Price  Low Price  \
1  MINDTREE     EQ 2017-05-16      509.45      507.00      515.85     506.00   
2  MINDTREE     EQ 2017-05-17      511.30      507.05      519.70     502.75   
3  MINDTREE     EQ 2017-05-18      506.35      502.00      516.75     486.00   
4  MINDTREE     EQ 2017-05-19      492.90      491.95      502.95     489.50   
5  MINDTREE     EQ 2017-05-22      497.50      499.95      511.85     497.10   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
1      511.00       511.30         512.69                 207011   
2      504.00       506.35         512.10                 391401   
3      487.50       492.90         505.60                 852336   
4      495.15       497.50         496.43                 352273   
5      501.55       502.35         504.78                 399974   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
1  1.061317e+08           5550            82090                   39.65   
2  2.004375e+08          13341           150098                   38.35   
3  4.309377e+08          18513           170919                   20.05   
4  1.748782e+08           8326            57853                   16.42   
5  2.018999e+08           9779           107089                   26.77   

   Index  Month  Year  Day_Perc_Change Trend  
1      1      5  2017         0.363137    NA  
2      2      5  2017        -0.968120    NA  
3      3      5  2017        -2.656265    NA  
4      4      5  2017         0.933252    NA  
5      5      5  2017         0.974874    NA

In [73]:
data.tail()

Symbol Series       Date  Prev Close  Open Price  High Price  \
489  MINDTREE     EQ 2019-05-07      981.75       982.8      985.00   
490  MINDTREE     EQ 2019-05-08      979.95       981.0      985.00   
491  MINDTREE     EQ 2019-05-09      980.50       979.6      983.15   
492  MINDTREE     EQ 2019-05-10      980.45       980.4      982.90   
493  MINDTREE     EQ 2019-05-13      975.35       973.4      983.00   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
489     975.00       981.0       979.95         979.77                1242934   
490     971.05       982.0       980.50         979.84                4301316   
491     976.10       980.9       980.45         980.02                3787159   
492     973.00       974.0       975.35         977.55                 215170   
493     970.30       981.0       979.10         980.16                 485941   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
489  1.217784e+09          10864           884988                   71.20   
490  4.214603e+09          13038          3851210                   89.54   
491  3.711482e+09           8159          3479880                   91.89   
492  2.103396e+08           5128            26734                   12.42   
493  4.762997e+08          14005           222691                   45.83   

     Index  Month  Year  Day_Perc_Change Trend  
489    489      5  2019        -0.183346    NA  
490    490      5  2019         0.056125    NA  
491    491      5  2019        -0.005099    NA  
492    492      5  2019        -0.520169    NA  
493    493      5  2019         0.384477    NA